app/com/linkedin/drelephant/tuning/ExecutionEngine.java

```diff
@@ -0,0 +1,83 @@
+package com.linkedin.drelephant.tuning;
+
+import com.avaje.ebean.ExpressionList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import models.AppResult;
+import models.JobDefinition;
+import models.JobExecution;
+import models.JobSuggestedParamSet;
+import models.JobSuggestedParamValue;
+import models.TuningAlgorithm;
+import models.TuningJobDefinition;
+import models.TuningParameter;
+import models.TuningParameterConstraint;
+
+
+/**
+ * Exeuction Engine is the interface of different execution engines. Currently it have two implementations
+ * MRExecutionEngine : Handle parameters releated to Map Reduce.
+ * SparkExecutionEngine : Handle parameters releated to Spark
+ */
+
+public interface ExecutionEngine {
+
+  /**
+   * This method is used to compute the values of derived parameters . These parameter values have not suggested by tuning algorithm
+   *
+   * @param derivedParameterList Derived Parameter List
+   * @param jobSuggestedParamValue Update job suggested param value with the derived parameter value list.
+   *
+   */
+  void computeValuesOfDerivedConfigurationParameters(List<TuningParameter> derivedParameterList,
+      List<JobSuggestedParamValue> jobSuggestedParamValue);
+
+  /**
+   * This method is used to get the pending JobSuggestedParamSet , param set which are not in FitnessCompute & discarded jobs
+   *
+   * @return Return pending jobs
+   */
+
+  ExpressionList<JobSuggestedParamSet> getPendingJobs();
+
+  /**
+   * This method is used to get jobs for which tuning enabled. Subtraction of this set with the above set
+   * gives us the jobs for which parameter have to be generated
+   *
+   * @return Return Tuning enabled jobs
+   */
+  ExpressionList<TuningJobDefinition> getTuningJobDefinitionsForParameterSuggestion();
+
+  /**
+   * Check if the suggested parameters violets any constraint for PSO .
+   */
+  Boolean isParamConstraintViolatedPSO(List<JobSuggestedParamValue> jobSuggestedParamValueList);
```

Comment:
It seems somewhat awkward to have an isParamConstraintViolated* for PSO, IPSO, and HBT. As we add more tuning methods, we would need to add more methods to this class. Could this be generalized, with an abstract class, which the different tuning types could inherit from?

Reply:
ok  so this is little complex. I have used bridge pattern to bridge ParamGenerator and execution engine as ParamGeneration would be different for different execution engines . If we use bridge pattern, then we have to go with the above way .  Now if we go with the suggested way , we will be having too many classes , possible n*m (where n is the execution engine types and m is algorithm type . We also may not have the implementation of that . So for e.g classes would be MRExecutionEnginePSO , MRExecutionEngineIPSO , SparkExecutionEnginePSO , SparkExecutionEngineIPSO , MRExecutionEngineHBT, SparkExecutionEngineHBT . Whenever we add , more tuning types or execution engine , we will have to create multiplication of classes . Well there is nothing wrong in creating those classes , but we will have lot more classes . 

Though I agree adding methods in class also may not be the best way . We are actually thinking to see other options (may be used generics ) , and may be next version will change , it to the suggested way (of having classes for all possible combination).

Reply:
That makes sense, a multiplicative number of classes would be a lot. If the implementations for each are different and independent, then it may still make sense to have separate classes. Generics sounds interesting.

Reply:
Yes, different PR sounds fine.

Reply:
https://jira01.corp.linkedin.com:8443/browse/DSS-18337